Mathlib/CategoryTheory/Limits/FormalCoproducts.lean

```diff
@@ -0,0 +1,426 @@
+  right_inv s := by ext <;> simp
+
+/-- `pullbackCone f g pb` is a pullback. -/
+def isLimitPullback : IsLimit (pullbackCone f g pb) := by
+  refine PullbackCone.IsLimit.mk
+    (fst := (pullbackCone f g pb).fst) (snd := (pullbackCone f g pb).snd) _
+    (fun s ↦ (homPullbackEquiv f g pb hpb s.pt).2 ⟨(s.fst, s.snd), s.condition⟩)
+    (fun s ↦ congrArg (·.1.fst)
+      ((homPullbackEquiv f g pb hpb s.pt).right_inv ⟨(s.fst, s.snd), s.condition⟩))
+    (fun s ↦ congrArg (·.1.snd)
+      ((homPullbackEquiv f g pb hpb s.pt).right_inv ⟨(s.fst, s.snd), s.condition⟩))
+    (fun s m h₁ h₂ ↦ ?_)
+  convert ((homPullbackEquiv f g pb hpb s.pt).left_inv m).symm using 3
+  rw [← h₁, ← h₂]; rfl
+
+-- Arguments cannot be inferred.
+include pb hpb in
+theorem hasPullback_of_PullbackCone : HasPullback f g :=
+  ⟨⟨⟨_, isLimitPullback f g pb hpb⟩⟩⟩
+
+omit pb
+variable [HasPullbacks C]
+
+instance : HasPullback f g :=
+  hasPullback_of_PullbackCone f g (fun _ ↦ pullback.cone _ _) (fun _ ↦ pullback.isLimit _ _)
+
+instance : HasPullbacks (FormalCoproduct.{w} C) :=
+  hasPullbacks_of_hasLimit_cospan _
+
+include pb
+
+/-- The arbitrary choice of pullback is isomorphic to the explicitly constructed pullback
+`pullbackCone f g pb`. -/
+noncomputable def pullbackIsoPullbackCone : pullback f g ≅ (pullbackCone f g pb).pt :=
+  limit.isoLimitCone ⟨_, isLimitPullback f g pb hpb⟩
+
+@[reassoc (attr := simp)] lemma pullbackIsoPullback_hom_fst :
+    (pullbackIsoPullbackCone f g pb hpb).hom ≫ (pullbackCone f g pb).fst = pullback.fst f g :=
+  limit.isoLimitCone_hom_π _ _
+
+@[reassoc (attr := simp)] lemma pullbackIsoPullback_inv_fst :
+    (pullbackIsoPullbackCone f g pb hpb).inv ≫ pullback.fst f g = (pullbackCone f g pb).fst :=
+  limit.isoLimitCone_inv_π _ _
+
+@[reassoc (attr := simp)] lemma pullbackIsoPullback_hom_snd :
+    (pullbackIsoPullbackCone f g pb hpb).hom ≫ (pullbackCone f g pb).snd = pullback.snd f g :=
+  limit.isoLimitCone_hom_π _ _
+
+@[reassoc (attr := simp)] lemma pullbackIsoPullback_inv_snd :
+    (pullbackIsoPullbackCone f g pb hpb).inv ≫ pullback.snd f g = (pullbackCone f g pb).snd :=
+  limit.isoLimitCone_inv_π _ _
+
+end Pullback
+
+
+noncomputable section HasCoproducts
+
+variable [HasCoproducts.{w} A] (C) (J : Type w) (f : J → FormalCoproduct.{w} C) (F : C ⥤ A)
+
+/-- A copresheaf valued in a category `A` with arbitrary coproducts, can be extended to the category
+of formal coproducts. -/
+@[simps] def eval : (C ⥤ A) ⥤ (FormalCoproduct.{w} C ⥤ A) where
+  obj F :=
+    { obj X := ∐ fun (i : X.I) ↦ F.obj (X.obj i)
+      map {X Y} f := Sigma.desc fun i ↦ F.map (f.φ i) ≫ Sigma.ι (F.obj ∘ Y.obj) (f.f i)
+      map_comp _ _ := Sigma.hom_ext _ _ (fun _ ↦ by simp [Sigma.ι_desc]) }
+  map α := { app f := Sigma.map fun i ↦ α.app (f.obj i) }
+
+/-- `eval(F)` restricted to the original category (via `of`) is the original copresheaf `F`. -/
+def evalOf : eval C A ⋙ (whiskeringLeft _ _ A).obj (of C) ≅ Functor.id (C ⥤ A) :=
```

Comment:
```suggestion
@[simps!]
def evalOf : eval C A ⋙ (whiskeringLeft _ _ A).obj (of C) ≅ Functor.id (C ⥤ A) :=
```
The generated lemmas look reasonable here IMO. Same for evalOpOf

Reply:
done